### ISSUE TITLE
Training with very small objects

### ISSUE THREAD

**Issue by james-nvidia** [NVIDIA/retinanet-examples#158]
_Originally posted by @VincentDuf in https://github.com/NVIDIA/retinanet-examples/issues/8#issuecomment-602027330_

Hi,
I have the same issue using the fine-tune method. I changed the learning rate twice (0.0005 and even 0.0001) but still after few ten's of thousand of iterations it prints "no detections" and I confirmed it after shutting the training down and try using "retinanet infer"...

I'm wondering if my objects aren't too small, I'm trying to detect aphids on plants (see image linked, [https://fromsmash.com/5_svMDKrmh-ct](url)) and I have a small dataset of 2840 images for training and 720 for validation. 
(2 precisions:
- I only labelled the aphids on the leaf not on the ground, there were too much of them
- I have 2 classes of aphids, green and pink, but on an image there are really similar so maybe I should merge them as one).
So do you think I should do more augmentations because Retinanet can handle that size of object of should move to segmentation algorithm because this is too small ? 

Thank you

**Comment by james-nvidia** [comment]
Those objects are very small. Can you try training again, but this edit `model.py` here:
https://github.com/NVIDIA/retinanet-examples/blob/master/retinanet/model.py#L29

Can you set
```
self.scales = [1, 2, 4]
```

**Please let me know if that works.**
If not, try changing these lines:
https://github.com/NVIDIA/retinanet-examples/blob/master/retinanet/box.py#L290
https://github.com/NVIDIA/retinanet-examples/blob/master/retinanet/box.py#L291
https://github.com/NVIDIA/retinanet-examples/blob/master/retinanet/box.py#L301

So that all of the `0.5` or `0.4` values become `0.2`.

**Comment by james-nvidia** [comment]
One more idea -- try this before changing the hardcoded values in `box.py`: Please also try padding your annotations by a few pixels (eg four)
So if your existing annotation is
```
x, y, w, h
```
Try training with new annotations as
```
x - 2, y - 2, w + 2, h + 2
```

**Comment by VincentDuf** [comment]
Ok thanks, so if I sum up, I should try:

1. training with modified _model.py_ file
2. training with padding annotations (with original or modified model.py ?)
3. training with modified _box.py_ (again with original model.py and annotations or with the changes ?)

Thanks

**Comment by james-nvidia** [comment]
Thinking about it, I'd try padding the annotations first, then try reducing `self.scales` in `model.py` (with the padded annotations.
Let me know how that goes before you try changing the IoU threshold in `box.py`.

**Comment by VincentDuf** [comment]
Hi,
I just finished with the `self.scales` modified training but it didn't help much (I only trained on 30.000 iterations as I'm fine-tuning and not full-training). Should I keep the learning rate equal to 0.0005 or should I try 0.0001 ? 

I'll add the padded annotations to see how it goes, and increase to 45k iterations.

**Comment by james-nvidia** [comment]
What focal and box loss are you seeing? While we're experimenting, I'd ramp your learning rate up as high as you can (eg 0.001) so that you can see if your network is learning.
The problem is that your targets aren't registering as foreground objects because the IoU between your targets and the anchors is below the threshold. We are trying three things:
1. Reducing the size of the anchors, which will increase the IoU with your targets.
2. Increasing the size of your targets (the padding), to also increase the IoU.
3. Decreasing the IoU threshold at which the network regards an object as foreground, so starts to predict it.

We will find a combination that works. Perhaps try with more padding (eg 10 pixels each side), show that it works, and then reduce the padding and see what effect that has.

What size is your original image? Are you using the default `--resize 800`? You could increase this (and also increase `--jitter` during training) to make your targets bigger with respect to the anchors.

**Comment by VincentDuf** [comment]
So far I see a low value for the first few iterations and then it falls to 0 (both bbox and focal loss). 
Okay so I just tried the 2 first option (padded + anchors scale), I'll now add the IoU threshold and increase the learning rate. 

My images are 1080p (1920x1080px), below that it was too hard to label the object manually. I don't use the resize option so it takes 1333p as a max from what I understand.
I didn't get what the jitter option would be useful for.. 

I keep on trying and I'll let you know how it goes. I'll increase the padded value to.

**Comment by james-nvidia** [comment]
Those losses are too low -- your targets are disappearing into the background.
Try adding `--jitter 1000 1200 --max-size 2000 --resize 1080` to your train command. This will resize your training images so that the short length is between 1000 and 1200 pixels and the long length is a maximum of 2000 pixels.

We want to see your losses greater than 0.1 for the first 30k iterations, otherwise it's not learning.

Have you labelled all of the aphids in the scene?

What padding did you use? Maybe add a little more until we see it learning properly.

**Comment by james-nvidia** [comment]
@VincentDuf As a reminder, make sure that you run `pip install -e .` from within the container, or rebuild the container, after you change the code.

**Comment by VincentDuf** [comment]
Ah yeah I forgot to rebuild the repo after the changes...I don't use the docker but still I didn't do it. 
Ok I'll add the options. 

I labelled the aphids that are on the plants only to create the association plant/aphid otherwise it would have been a titan work to label all of them especially the one on the ground. More over my interest is to detect aphids on the plant (the part on the ground isn't really representative of a real density of aphids around plants).

I applied the padding twice so far so it's already [x–4,y-4,w+8,h+8].

I rebuild the whole thing and launch another try.

**Comment by james-nvidia** [comment]
"Detect object A only when it is on object B" is a hard training signal -- the network is set up for "detect objects A and B" and then post processing to detect an overlap.

Another thing to consider is to crop your training images to the area that you've fully labelled (for now), then the task becomes just object detection. Remember to update the values in `--jitter 1000 1200 --max-size 2000 --resize 1080` to make sure they make sense with your smaller dataset.

I'd try all of these things (smaller anchors, padding, crop your image to the area labelled) before you adjust the IoU threshold.

**Comment by VincentDuf** [comment]
Oh okay I supposed falsely that when training a network the "context" mattered, like in my case where an aphid on a plant is considered so because it's essentially a grey dot in a green area (gross idea) and a grey dot in a brow area (the ground) wouldn't be consider as an aphid..

Ok yeah cropping the images is a good idea, I'm working in parallel on a simple plant detector with classic image processing techniques and that could be a part of 2 stages detector. I'll dig into that.
It will probably be hard to delete all the non labelled object from the image but I can definitely reduce the number.

Okay thank you very much for your extensive help. I learned a lot 👍

**Comment by james-nvidia** [comment]
One last thought: One of the 80 COCO classes is "potted plant". You can look at some examples here: http://cocodataset.org/#explore

We provide trained models (trained on COCO) for this network, under the 'release' tab. You might be able to use that network to detect your plants, then crop your images in an automated fashion.

I'll close this now, but please open a new issue if you have problems.

**Comment by VincentDuf** [comment]
Oh so you would crop the images using the network then augment the dataset and then train a new one using retinanet to detect aphids on that plant ?
I'll have a look to how I can do that.
Thank you again I'll let you know how things go with the training.

**Comment by VincentDuf** [comment]
Hi,
I'm facing an error I don't really get what's the problem:
```
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/numpy/core/function_base.py", line 117, in linspace
    num = operator.index(num)
TypeError: 'numpy.float64' object cannot be interpreted as an integer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/greenshield/.virtualenvs/cv/bin/retinanet", line 11, in <module>
    load_entry_point('retinanet==0.1', 'console_scripts', 'retinanet')()
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/main.py", line 181, in main
    worker(0, args, 1, model, state)
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/main.py", line 127, in worker
    metrics_url=args.post_metrics, logdir=args.logdir, verbose=(rank == 0))
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/train.py", line 154, in train
    mixed_precision=mixed_precision, is_master=is_master, world=world, use_dali=use_dali, is_validation=True, verbose=False)
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/infer.py", line 141, in infer
    coco_eval = COCOeval(data_iterator.coco, coco_pred, 'bbox')
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/pycocotools/cocoeval.py", line 76, in __init__
    self.params = Params(iouType=iouType) # parameters
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/pycocotools/cocoeval.py", line 527, in __init__
    self.setDetParams()
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/pycocotools/cocoeval.py", line 507, in setDetParams
    self.iouThrs = np.linspace(.5, 0.95, np.round((0.95 - .5) / .05) + 1, endpoint=True)
  File "<__array_function__ internals>", line 6, in linspace
File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/numpy/core/function_base.py", line 121, in linspace
    .format(type(num)))
TypeError: object of type <class 'numpy.float64'> cannot be safely interpreted as an integer.
```

I didn't touch anything from the database and it's the second time it occurs at the exact same number of iteration. 
If you want the full log file I have it.

**Comment by james-nvidia** [comment]
That error is coming from `pycocotools` -- my guess is it doesn't like something in your dataset.
Can you see if you can find an error in your data format?

If not, please download the latest version of this repo (you are on version 0.1, latest is 0.2.1) and then open a new issue if you're sure it's not a dataset problem?

**Comment by VincentDuf** [comment]
Hey,
I found the issue, it was caused by numpy, the version 1.18 isn't compatible with pycocotools, here is the thread that helped me:
https://stackoverflow.com/questions/59493606/pytorch-and-tensorflow-object-detection-evaluate-object-of-type-class-nump

It is finally learning and I even have those printed information after a first epoch:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.094
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.238
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.053
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.094
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.020
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.098
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.126
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.126
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
```
Can't tell what it really means but it looks like the process is finally running properly.

**Comment by VincentDuf** [comment]
Hi,
I'm facing a space issue, what's the procedure to correct this ? 
(To be noted I have more images than previously..)
```
Traceback (most recent call last):
  File "/home/greenshield/.virtualenvs/cv/bin/retinanet", line 11, in <module>
    load_entry_point('retinanet==0.1', 'console_scripts', 'retinanet')()
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/main.py", line 181, in main
    worker(0, args, 1, model, state)
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/main.py", line 127, in worker
    metrics_url=args.post_metrics, logdir=args.logdir, verbose=(rank == 0))
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/train.py", line 80, in train
    cls_loss, box_loss = model([data, target])
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/apex/amp/_initialize.py", line 197, in new_fwd
    **applier(kwargs, input_caster))
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/model.py", line 111, in forward
    return self._compute_loss(x, cls_heads, box_heads, targets.float())
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/model.py", line 158, in _compute_loss
    cls_loss = self.cls_criterion(cls_head, cls_target)
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/retinanet-0.1-py3.5-linux-x86_64.egg/retinanet/loss.py", line 15, in forward
    ce = F.binary_cross_entropy_with_logits(pred_logits, target, reduction='none')
  File "/home/greenshield/.virtualenvs/cv/lib/python3.5/site-packages/torch/nn/functional.py", line 2126, in binary_cross_entropy_with_logits
    return torch.binary_cross_entropy_with_logits(input, target, weight, pos_weight, reduction_enum)
RuntimeError: CUDA out of memory. Tried to allocate 52.00 MiB (GPU 0; 7.94 GiB total capacity; 7.27 GiB already allocated; 32.56 MiB free; 7.44 GiB reserved in total by PyTorch)
```

Thanks

**Comment by james-nvidia** [comment]
What batch size are you using? Try reducing it to the number of GPUs you are using, so if one GPU, try `--batch 1`

**Comment by VincentDuf** [comment]
Thank you it worked so far. 
Another question: I have now 4200 images for the training of one classe only do you think I can try a full training instead of fine-tuning the pre-trained model ? 
Thanks

**Comment by james-nvidia** [comment]
Both would work, but I expect fine-tuning to work better. Try both if you have the time / compute resources.

**Comment by VincentDuf** [comment]
Hey,
I just ran a complete fine-tune 90K iteration training. The losses look right (they are all > 0.1 for the whole process). 
At the end I have these annotations:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.264
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.665
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.138
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.264
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.029
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.217
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.372
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.372
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
```

How should I interprete those number ? 

Thanks

**Comment by james-nvidia** [comment]
This explains the pycocotools output:
https://medium.com/@timothycarlen/understanding-the-map-evaluation-metric-for-object-detection-a07fe6962cf3

What do you mean about 3/4 times more annotations? In your training set?

**Comment by VincentDuf** [comment]
> 
> 
> This explains the pycocotools output:
> https://medium.com/@timothycarlen/understanding-the-map-evaluation-metric-for-object-detection-a07fe6962cf3
> 
> What do you mean about 3/4 times more annotations? In your training set?

No, I meant the network annotates 3 to 4 times more objects than the validation dataset. It finds way more objects than I want him to.

Should I reduce first the number of training iteration or is it more linked to the NonMaxSuppression parameter (how to change it in retinanet?) or the IoU levels you mentionned ?

Thanks

**Comment by james-nvidia** [comment]
How are you plotting your results images? Only plot those with score about a threshold. Increase that threshold to get fewer detections.
It looks like you threshold is very low. What is it at the moment?

**Comment by VincentDuf** [comment]
Oh okay I thought the output already took the threshold in account. I'll have a look, I'll let you know.
I'm using this script btw to plot the results:
```
import json
import cv2
import matplotlib.pyplot as plt
import numpy as np
import os

def getImageId(name,data):
    images = data['images']
    for image_data in images:
        if image_data['file_name']==name:
            return image_data['id']

        
def showBoxes(image_path,annotation_file):
    image = cv2.imread(image_path)
    (h1, w1) = image.shape[:2]	
    center = (w1 / 2, h1 / 2)
    with open(annotation_file) as f:
        data = json.load(f)
    image_id = getImageId(image_path[image_path.rfind('/')+1:],data)
    boxes = []
    for annot in data['annotations']:
        if annot['image_id']==image_id:
            boxes.append((annot['category_id'],annot['bbox']))
    cloneImg = image.copy()
    for box in boxes:
        (_,(x,y,w,h)) = box
        cv2.rectangle(cloneImg,(int(x),int(y)),(int(x+w),int(y+h)),(255,0,0),1)
    cloneImg = cloneImg[:-50,int(center[0]-300):int(center[0]+550)]
    return cloneImg[:,:,::-1]


listfile = os.listdir('validation/')
del listfile[listfile.index('annotations.json')]


for name in listfile:
	img = showBoxes('validation/'+name,'validation/annotations.json')
	img_trained = showBoxes('validation/'+name,'valid_detect.json')
	buff = np.concatenate((img,img_trained),axis=1)		
	#cv2.imshow('Comparaison',buff)
	#cv2.waitKey(0)
	cv2.imwrite('compare/'+name,buff)
	print(name)

print('done')

```

It's coming from this repo: https://github.com/joheras/CLoDSA

**Comment by VincentDuf** [comment]
> 
> 
> How are you plotting your results images? Only plot those with score about a threshold. Increase that threshold to get fewer detections.
> It looks like you threshold is very low. What is it at the moment?

Okay I modified my script to take into account the score and it looks better. To find the best threshold I assume I need to use the COCO mAP metrics right ? I'll implement these using the article on Medium that you provided.

**Comment by VincentDuf** [comment]
Hi,
I have a question I can't find the answer to: when I use the mAP metrics I have some numbers out, but it looks like it only takes in account IoU and mAP. 
But with what threshold of confidence (the 'score') are they calculated on? 

I mean the mAP of the network with a threshold of 5% won't be the same with a threshold of 80% right ? I can't get what is the relation between mAP/IoU and the "score" of the boxes (and so the threshold I can choose to filter).  

My idea would be to evaluate the mAP of the network for different threshold on the score to find the best threshold. Does it make sense ? 

Thanks

**Comment by james-nvidia** [comment]
There's a whole bunch of discussion about the best way to report metrics for object detection. I personally favour a precision/recall curve. In this repo we use `pycocotools` mAP because it's the standard for COCO object detection. You might want to use your own metric (F1, F2, etc) if you have a specific use case.
Alternatively there's a lot more you can do using pycocotools. You could edit `infer.py` and change the call there.

**Comment by VincentDuf** [comment]
Okay I rephrase my question, it wasn't clear. 
When the mAP metrics are calculated, they are using the model trained but there is no restriction on the "score" right ? So the mAP are "valid" for the boxes with the confidence score > 0 only. 

My question was then: is it possible to re-compute the mAP metrics but incorporating a threshold on the confidence score ? For example I want to test the mAP for boxes with a score > 0.15, is it possible (by modifying the infer.py for example) ? 

Thanks

**Comment by VincentDuf** [comment]
Hi,
I just noticed a weird behaviour using the "infer.py" script to evaluate the model precision.
When I use the `--with-dali` option, the time/img decreased significantely (by 2 at least) and the precision metrics increase a little.

Here are the two results metrics, using the exact same model and images/annotations but one with pytorch loader and the other with dali:

```
retinanet infer rn50fpn/rn50fpn_crop.pth --images validation_crop/ --annotations validation_crop/annotations.json
Loading model from rn50fpn_crop.pth...
     model: RetinaNet
  backbone: ResNet50FPN
   classes: 1, anchors: 9
Preparing dataset...
    loader: pytorch
    resize: 800, max: 1333
   backend: pytorch
    device: 1 gpu
     batch: 2, precision: mixed
Running inference...
[90/90] 0.347s/2-batch (fw: 0.344s), 5.8 im/s
Gathering results...
Writing detections.json...
Evaluating model...
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.371
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.829
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.251
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.371
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.034
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.280
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.472
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.472
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
```

```
$ retinanet infer rn50fpn/rn50fpn_crop.pth --images validation_crop/ --annotations validation_crop/annotations.json --with-dali

Loading model from rn50fpn_crop.pth...
     model: RetinaNet
  backbone: ResNet50FPN
   classes: 1, anchors: 9
Preparing dataset...
    loader: dali
    resize: 800, max: 1333
   backend: pytorch
    device: 1 gpu
     batch: 2, precision: mixed
Running inference...
[90/90] 0.714s/2-batch (fw: 0.712s), 2.8 im/s
Gathering results...
Writing detections.json...
Evaluating model...
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.398
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.863
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.278
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.398
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.032
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.293
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.512
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.512
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
```

Any ideas why it does this and what are the "real" metrics ?

Thanks

**Comment by james-nvidia** [comment]
The difference in mAP is explained by the two dataloaders behaving slightly differently. Both scores are valid.
DALI would normally increase your inference speed, but it may be that your dataset is so small that the overhead of setting up DALI outweighs the increase in data loading speed.
What GPU are you using? 
Have you considered using the new MobileNetV2FPN backbone? It should run more quickly.

**Comment by VincentDuf** [comment]
Ok it's good to know thanks for the explanation, it just feels weird that the metrics can be impacted by the dataloader..

Here is my config:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64.00    Driver Version: 440.64.00    CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Quadro M4000        Off  | 00000000:01:00.0 Off |                  N/A |
| 46%   31C    P8    10W / 120W |     24MiB /  8125MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1221      G   /usr/lib/xorg/Xorg                            21MiB |
+-----------------------------------------------------------------------------+

```

I didn't see you added MobileNet, I'll try it right away ! 

Thanks

**Comment by VincentDuf** [comment]
Hey,
I'd like to know wether the focal loss is calculated on the test dataset or the validation one ? 

I also would like to modify this focal loss to tune it for my case. I want to add a term taking into account the precision of the center of my predicted bounding boxes. In what file should I add this term ? 

Thank you
